core/templates/dev/head/pages/exploration_editor/editor_tab/StateSolution.js

```diff
@@ -0,0 +1,224 @@
+            $scope.EXPLANATION_FORM_SCHEMA = {
+              type: 'html',
+              ui_config: {}
+            };
+
+            $scope.submitAnswer = function(answer) {
+              $scope.correctAnswer = answer;
+            };
+
+            $scope.tmpSolution = {};
+            $scope.tmpSolution.answerIsExclusive = false;
+            $scope.tmpSolution.correctAnswer = '';
+            $scope.tmpSolution.explanation = '';
+
+            $scope.addSolutionForm = {};
+
+            $scope.saveSolution = function() {
+              // Close the modal and save it afterwards.
+              var answer = $scope.correctAnswer;
+              if (!interactionHtml) {
+                answer = $scope.tmpSolution.correctAnswer;
+              }
+              var answerObject = (
+                StateSolutionHelperService.getCorrectAnswerObject(answer,
+                  $scope.objectType));
+              $modalInstance.close({
+                solution: SolutionObjectFactory.createNew(
+                  $scope.tmpSolution.answerIsExclusive,
+                  answerObject,
+                  $scope.tmpSolution.explanation)
+              });
+            };
+
+            $scope.cancel = function() {
+              $modalInstance.dismiss('cancel');
+              alertsService.clearWarnings();
+            };
+          }
+        ]
+      }).result.then(function(result) {
+        var correctAnswer = result.solution.correctAnswer;
+        var currentStateName = editorContextService.getActiveStateName();
+        var state = explorationStatesService.getState(currentStateName);
+        try {
+          StateSolutionHelperService.verifyAndSaveAnswer(
+            explorationContextService.getExplorationId(),
+            state,
+            correctAnswer);
+        } catch (e) {
+          alertsService.addInfoMessage('That solution was invalid!');
+          $scope.openAddSolutionModal();
+        }
+        stateSolutionService.displayed = result.solution;
+      });
+    };
+
+    $scope.deleteSolution = function(evt) {
+      evt.stopPropagation();
+
+      alertsService.clearWarnings();
+      $modal.open({
+        templateUrl: UrlInterpolationService.getDirectiveTemplateUrl(
+          '/pages/exploration_editor/editor_tab/deleteSolutionModal.html'),
+        backdrop: true,
+        controller: [
+          '$scope', '$modalInstance',
+          function($scope, $modalInstance) {
+            $scope.reallyDelete = function() {
+              $modalInstance.close();
+            };
+
+            $scope.cancel = function() {
+              $modalInstance.dismiss('cancel');
+              alertsService.clearWarnings();
+            };
+          }
+        ]
+      }).result.then(function() {
+        stateSolutionService.displayed = null;
+        stateSolutionService.saveDisplayedValue();
+        explorationStatesService.getState(
+          editorContextService.getActiveStateName()
+        ).interaction.markSolutionAsInvalid();
```

Comment:
Seems weird that you do this in delete but not in add. Is there a central location that all the "modify solution" paths go through where you can add an "update validity" step? Otherwise this is going to be impossible to maintain.